Add Galaxy Notification System

davelopez opened the pull request:
Builds on (and supersedes) the work started in #14041

This is an overhaul of the originally designed `Notification System` to support more complex scenarios and features described in #1023 and #2324

# Features

### Notification Scheduling

Notifications can be scheduled to be displayed to the users on a future date by specifying a `publication_time`.

### Notification Expiration

Notifications will have a default expiration date and will be removed permanently from the database regularly after their expiration date is over. Either the `cleanup` scripts or a cron-like celery task can regularly remove expired notifications to avoid bloating the DB.

### Notification Broadcasting

A special type of notification meant to support server-wide announcements. These notifications will be usually treated differently than regular notifications and will be displayed in the main view as banners or alerts for every user (authenticated or anonymous).
Some use cases for these notifications can be:

- Server scheduled maintenance notifications.
- Server high load warnings.
- Event announcements.
- User agreements (cookies, privacy policy, etc.).

### Structured Notifications

Notifications can have different categories and _variants_ and their content (JSON) is well-defined depending on the category. In general, the notification may just contain the necessary metadata to build the actual notification message in the client using the appropriate localization and display style.

#### Possible notification categories:

- `message`: A direct message to the user.
- `new_shared_item`: new resource (history, workflow, visualization or page) shared with the user, can contain the link to the shared resource.
- `workflow_execution_completed`: There is no hook yet for this event, but it will be possible in the future.
- More ideas here...

### More control by the user

#### User Preferences

The user can decide **what** and **how** notifications will be displayed by opting out of specific notification categories or channels.
All notifications will be displayed in a dedicated panel in the Galaxy UI unless completely opt-out, in addition, the user can decide on which channels the notification can be sent.

Possible notifications channels un/selectable by the user for each notification category:

- push: This will display a web push notification
- email: The notification can be optionally sent by email (not included in this PR)
- matrix: Direct message via Matrix.org (not included in this PR)

These notification channels are just an idea for now and their implementations may or may not be included in this PR.

#### Favorites

~~The user may mark some notifications as `favorite`. This will prevent the notification from automatically expiring and offers more filtering options. This might be interesting, for example, to save a notification with useful information or links for future use.~~
EDIT: This has been removed.

#### Deletion

The user can always manually delete the notification without having to wait for the expiration date.

# Backend Changes

## Database models

The database models have been extended to include more information and support the above features:

```mermaid
erDiagram
    NOTIFICATION ||--o{ USER_NOTIFICATION_ASSOCIATION : notifies_users
    NOTIFICATION {
        int id PK
        datetime create_time "The creation date."
        datetime update_time "The last time the notification content was updated."
        datetime publication_time "The date of publication, can be a future date to allow scheduling."
        datetime expiration_time "The expiration date, expired notifications will be permanently removed from DB regularly."
        string source "Who (or what) generated the notification."
        string category "Category of the notification, defines its contents. Used for filtering, un/subscribing, etc."
        string variant "Defines the 'importance' of the notification. Used for filtering, highlight rendering, etc."
        json content "Structured content in JSON. Depending on the category of the notification."

    }
    USER_NOTIFICATION_ASSOCIATION {
        int id PK
        int user_id FK
        int notification_id FK
        datetime seen_time "Date when the user 'opened' the notification (or null if unread)."
        bool deleted "Marking a notification as deleted will hide it from the user until it expires. Can be undeleted."
    }
```

## API

The rest API has been also extended to support the features and use cases described above.

Access the interactive API documentation here: http://127.0.0.1:8080/api/docs#/notifications

![image](https://github.com/galaxyproject/galaxy/assets/46503462/34d12a9c-66c3-4e7c-9314-2a419bbb8655)


## TODO
- [x] Use the user preferences to skip notification creation
   - [ ] Create associations in an external task when targeting a potentially large number of users? Let's see how it behaves and then follow-up in a different PR.
- [x] Add expiration/deletion cleanup mechanism
- [x] Add notification hooks for shared items (histories, workflows, etc.)
- [x] Add config option to globally enable/disable notifications
- [x] Add test coverage

# Frontend Changes

There are two main kinds of notifications, *User notifications* and *server-wide (or broadcast) notifications*.

## User notifications

User notifications are specific for each user and can be marked as read or deleted by them. In this PR two sub-types of user notifications are implemented, **Shared Items** and general **Message** notifications.

![image](https://github.com/galaxyproject/galaxy/assets/46503462/3b6d0cd4-cde1-4663-ac3f-d0de1502a2d3)

Notification preferences allow you to subscribe/unsubscribe from certain types of User Notifications and choose through which channels you want to be notified (currently only the *Push* notification channel is available others will be added in the future).

![image](https://github.com/galaxyproject/galaxy/assets/46503462/2582ad4a-d781-4cff-bc2f-3321bc498fa7)

## Broadcast notifications (server-wide)

There are again two sub-types of broadcasted notifications, depending on the *variant*. "regular" broadcast (info and warning variant) can be used to show Terms of use, Cookies policies, and all sorts of server-wide messages, while **urgent** variants are a bit more intrusive and require a bit more attention from the user.

### Regular broadcasts example
![Screenshot from 2023-06-16 16-04-17](https://github.com/galaxyproject/galaxy/assets/46503462/634ad9d8-78f3-4f31-94cc-2fcac7b6e076)

### Urgent variant example

![image](https://github.com/galaxyproject/galaxy/assets/46503462/12a154f3-ac41-4996-98af-e477b3dc1c24)


## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

shivani-sangwan committed:
Add Notification System API Database model

shivani-sangwan committed:
Adding mapping tests for Notification and UserNotificationAssociation

shivani-sangwan committed:
Adding fxitures and mapper tests for Notification and UserNotification

shivani-sangwan committed:
Adding API module for Notification Manager

shivani-sangwan committed:
Adding notification module in API and registering NotificationManager in structured_app.py

shivani-sangwan committed:
Adding API tests for Notifications

mvdbeek committed:
Move notiifcation api tests to lib/galaxy_test/api

shivani-sangwan committed:
Adding Payload Descriptor for Notification in Schema

shivani-sangwan committed:
Adding notification annotation response in API

mvdbeek committed:
Remove unique constraint on notification message text

mvdbeek committed:
Fix offset logic in NotificationManager index

mvdbeek committed:
Fix usage of schema models

shivani-sangwan committed:
Adding NotificationRequestModel in schema

shivani-sangwan committed:
Adding unit tests for notifications API

shivani-sangwan committed:
Adding UI component for notification display

commit 2a4e8575980f8bb736c87d298ca89b33dce2b8f2
Author: Shivani Sangwan <shivanisangwan31@gmail.com>
Date:   Tue Aug 9 19:32:34 2022 +0530

    Adding UI for notification list display

shivani-sangwan committed:
Adding user notification association while creating notifications

davelopez committed:
Basic fixes and cleanups

- Fix vue-notification-bell client dependency installation
- Fix linting issues
- Refactor rename some variables and types

davelopez committed:
Expand notification related tables

This will provide more control and flexibility to support different kinds of notifications.

davelopez committed:
Refactor NotificationManager tests

davelopez committed:
Make notification database models more flexible

davelopez committed:
Move and extend notification schema models

- Move models to schema/notifications.py
- Extend the models to increase flexibility and usability of the API

davelopez committed:
Update NotificationManager

Add `create_notification_for_users` function and `get_user_notifications`.

davelopez committed:
Add NotificationService scaffolding

davelopez committed:
Add FastAPI endpoints for notifications

davelopez committed:
Add broadcasted notifications support

davelopez committed:
Implement notification status endpoint

davelopez committed:
Make `since` date required for requesting status

mvdbeek commented on lib/galaxy/schema/notifications.py:
```diff
@@ -0,0 +1,256 @@
+    new_history_shared = "new_history_shared"
+    new_workflow_shared = "new_workflow_shared"
+    new_page_shared = "new_page_shared"
+    new_visualization_shared = "new_page_shared"
```
```suggestion
    new_visualization_shared = "new_visualization_shared"
```
?

davelopez replied:
Ups copy&paste is ruled by the devil :sweat_smile: Thanks!

mvdbeek commented on lib/galaxy/schema/notifications.py:
```diff
@@ -0,0 +1,256 @@
+    new_workflow_shared = "new_workflow_shared"
+    new_page_shared = "new_page_shared"
+    new_visualization_shared = "new_page_shared"
+    workflow_execution_completed = "workflow_execution_completed"
```
```suggestion
    workflow_execution_completed = "workflow_execution_completed"
```
😍 I'd love to build the backend for that, just be aware we don't have a way to hook into that yet

davelopez replied:
Ahh good to know, those categories are just potential ideas for the moment, I'll disable that one for now.

jmchilton replied:
near_quota_limit should also be disabled for now I assume (push back if I'm wrong). We present a quota bar to the user in the UI - the client could just look at this and decide what to do - rather than having to check and fire this every time we write any object in Galaxy. 

This is overall wonderful. I'd love to see the notifications implemented - the sharing stuff seems like a good start. There is also a job completion -> email hook in Galaxy that you can probably piggy back off of to build another useful thing pretty quickly.

davelopez replied:
Thank you! That makes a huge lot of sense @jmchilton!

I'll look into the job completion hook :+1:

mvdbeek commented on lib/galaxy/schema/notifications.py:
```diff
@@ -0,0 +1,256 @@
+
+class ActionLink(Model):
+    action_name: str
+    link: AnyUrl
```
Can we not do urls, but return entity ids, so the client can build whatever operation is appropriate for an entity ?

So if it's a workflow_id, you could view, edit, run, download etc ... ?

Take a look at `WarningReason`, `FailureReason`, `CancelReason` ? That way we also don't need to hook into the backend routing for generating routes

davelopez replied:
Sure, that was the idea. Have a look at `WorkflowExecutionCompletedNotificationContent` below.

This `ActionLink` is only used in *Broadcasted notifications*, which are generic, and this is a naive attempt to support "Actions" or "Buttons" that can register specific actions.

For example, imagine creating a survey that gets displayed to every user (as a broadcasted notification), and there are a couple of options, those options could be links to a potential API endpoint (inside or outside Galaxy) that register the response. 
Or another example is when you click "Agree" on a privacy policy or cookie agreement. There would be an `agree` action link that points somewhere to read the terms, etc... But maybe it is too generic... :thinking: 

Those action links are supposed to be written by the admin or the person in charge of creating those special notifications, so they are not generated by Galaxy.

mvdbeek replied:
Ah, great, for outside links that makes a lot of sense.

davelopez replied:
Ohh wait, I see what you mean, the `share_link` below should go :sweat_smile:

davelopez committed:
Fix typo in enumerate value

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

jdavcs commented on lib/galaxy/model/migrations/alembic/versions_gxy/f6fb225c05ab_add_notification_table.py:
```diff
@@ -0,0 +1,41 @@
+"""Add Notification Table
```
Could you, please, move the revision script into its own commit? I haven't seen a case (yet) when we'd need the revision commit without the corresponding model change commit, but that was one of the "lessons learned" items in a 2014 PyCon Alembic talk, and I do think it's a good approach (if anything, that's more granular commits, which may be convenient).

jdavcs commented on test/unit/data/model/mapping/test_gxy_model_mapping.py:
```diff
@@ -1317,6 +1317,84 @@ def test_relationships(self, session, cls_, group, role):
             assert stored_obj.role.id == role.id
 
 
```
We are dropping all mapping tests. It's not merged yet only because it got stuck in a large SA2.0 PR which is not yet ready for review. There's a more detailed explanation in the backend meeting minutes from 1/31/23 (see  bullet point 1.3). So this file will be dropped. (sorry for any wasted time - I should've dropped those already)

davelopez replied:
No worries! I had a vague feeling/memory this was the case but since the tests were already there in the original PR... I didn't think of it too much :man_shrugging: 

I'll drop them :+1: Thank you!

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -2457,6 +2458,46 @@ def __init__(self, user, group):
+    expiration_time = Column(DateTime, default=now() + timedelta(days=30 * 6))
+    source = Column(String(32), index=True)
+    category = Column(String(64), index=True)
+    variant = Column(String(16), index=True)
```
maybe add very brief inline comments (on the right?) explaining what there are? Like what you have in the data models table in the description of this PR. That would be very helpful for anyone reading the model definition code down the road.

jdavcs commented on lib/galaxy/model/__init__.py:
```diff
@@ -2457,6 +2458,46 @@ def __init__(self, user, group):
+    expiration_time = Column(DateTime, default=now() + timedelta(days=30 * 6))
+    source = Column(String(32), index=True)
+    category = Column(String(64), index=True)
+    variant = Column(String(16), index=True)
```
Does the `variant` field imply a scale? ("Defines the 'importance' of the notification.") If so, maybe use an integer instead? Not only will that give us more flexibility (avoid the need for "high" vs. "somewhat high" vs. "very very high", etc.), we also get natural ordering.

davelopez replied:
The idea was more along the line of something like "info", "warning", "urgent", etc. that can be used to render or highlight the notifications differently (colors, icons, etc.). I think the ordering should be always by date, but this `variant` can be used for filtering, i.e. display only "urgent" notifications.
Of course, these enums can be mapped as integers in the DB if this is more desirable, I just thought they are more readable as strings, but I'm happy to change that if this is more optimum :)

jdavcs replied:
Ah, I see. Those values are more informative than a simple integer scale - let's keep them as strings then.
With regard to the database, it's all the same here. Thanks for the explanation!

jdavcs commented on lib/galaxy/model/unittest_utils/gxy_model_fixtures.py:
```diff
@@ -895,6 +895,17 @@ def make_instance(*args, **kwds):
     return make_instance
 
 
```
Both fixtures can be dropped (together with the mapping tests) - again, sorry about that!

davelopez committed:
Update schema models

davelopez committed:
Add some code doc

davelopez committed:
Add Notification System database migration script

davelopez committed:
Drop unused notification fixtures

davelopez committed:
Implement get_user_notification

davelopez committed:
Implement notification updates

davelopez committed:
Refactor schema models

davelopez committed:
Implement notification preferences

davelopez committed:
Refactor imports in NotificationManager

davelopez committed:
Add single broadcast notification GET endpoint

davelopez committed:
Add some more basic tests

davelopez committed:
Update client API schema

davelopez committed:
Set ObjectNotFound message in service layer

So the manager does not have to handle any kind of ID encoding.

davelopez committed:
Allow to include or exclude expired notifications

davelopez committed:
Add function to cleanup expired notifications

davelopez committed:
Refactor Notifications manager

Use `active_only` for filtering published and not expired notifications.

davelopez committed:
Add notification hook for shared items

davelopez committed:
Refactor sharable item notifications

davelopez committed:
Refactor recipient resolver and add some tests

davelopez committed:
Refactor NotificationManager tests

davelopez committed:
Update client API schema

davelopez committed:
Add config option to enable/disable notifications

davelopez committed:
Add integration tests

Replaces API tests since Notifications are an optional feature

davelopez committed:
Exclude deleted notifications from queries

And small refactor

davelopez committed:
Refactor and increase test coverage

davelopez committed:
Check user preferences before sending notification

davelopez committed:
Refactor notification preferences + tests

davelopez committed:
Increase unit and integration tests coverage

davelopez committed:
Add periodic task to cleanup expired notifications

davelopez committed:
Return inactive bcast notifications for admins

davelopez committed:
Fix mypy

davelopez committed:
Include updated notifications on `since` request

itisAliRH committed:
add utility `mergeObjectListsById` with unit tests

itisAliRH committed:
create component `AsyncButton`

itisAliRH committed:
create notifications services

itisAliRH committed:
create notifications store

itisAliRH committed:
update notifications list: add icon, unread status, action buttons and filters

itisAliRH committed:
register notifications pulling in the app entry

itisAliRH committed:
update notifications router usage

davelopez committed:
Increase API documentation and improve schema

itisAliRH committed:
create `NotificationsBell` component

itisAliRH committed:
add indicator, tooltip placement and object icon to `ActivityItem.vue`

itisAliRH committed:
add `NotificationsBell` to `ActivityBar` and `Masthead`

itisAliRH committed:
add `slot` and disable on loading to `AsyncButton` component

itisAliRH committed:
create user notifications service

itisAliRH committed:
create user notifications preferences

itisAliRH committed:
add `user/notifications` client route

itisAliRH committed:
add new `Notifications Preferences` to user preferences

itisAliRH committed:
fix masthead showing notifications bell logic

itisAliRH committed:
refactor `NotificationsBell.vue`

dannon committed:
remove `vue-notification-bell` package and its component

itisAliRH committed:
fix `utils.test.js`: delete unused var

itisAliRH committed:
run prettier

itisAliRH committed:
create `pushNotifications.ts` util

itisAliRH committed:
refactor `NotificationsPreferences.vue`: use pushNotifications util and add alerts

itisAliRH committed:
add `NotificationsPreferences` to `NotificationsList.vue`

itisAliRH committed:
use `UserNotificationListResponse` type for `notifications` in `notificationsStore`

itisAliRH committed:
create `broadcasts.service.ts`

itisAliRH committed:
crate `broadcastsStore.ts`

itisAliRH committed:
create `BroadcastsOverlay.vue`

itisAliRH committed:
register `broadcasts` pulling and use `BroadcastsOverlay` in the app entry

davelopez committed:
Set default notification status polling delay to 5s

davelopez committed:
Unify notification status polling

davelopez committed:
Handle dismissed broadcast messages

itisAliRH committed:
add `title` with popover to `AsyncButton` and change default size from `sm` to `md`

itisAliRH committed:
create `NotificationActions` component

itisAliRH committed:
create `MessageNotification` component to render `message` category

itisAliRH committed:
create `SharedItemNotification` component to render `new_shared_item` category

itisAliRH committed:
update `NotificationsList` to use new components

davelopez committed:
Notify user only once on item sharing

davelopez committed:
Fail requests early if notifications are disabled

itisAliRH committed:
check in ui if notifications system is not enabled

itisAliRH committed:
cleanup notifications components and improving ui

davelopez committed:
Small refactorings and cleanups

itisAliRH committed:
refactoring notifications routes and files

itisAliRH committed:
use markdown to render broadcast content and fix styles

itisAliRH committed:
fix notifications preferences style and some other cleanups

davelopez committed:
Additional refactorings

- Generic NotificationItem component
- Add markdown to notification message

itisAliRH committed:
create notifications test utilities to be used in tests

itisAliRH committed:
add ids to notifications list to make it easier to test

itisAliRH committed:
add tests for NotificationsList component

itisAliRH committed:
add ids to NotificationsActions for tests

itisAliRH committed:
add tests for NotificationActions

itisAliRH committed:
add tests for notification categories

itisAliRH committed:
fix Categories test types

itisAliRH committed:
disable notifications bell for anonymous user

itisAliRH committed:
change urgent broadcasts position to the top and drop the dark background overlay

itisAliRH committed:
remove notifications favorite button from NotificationActions and its test

itisAliRH committed:
remove notifications favorite filter from NotificationsList and notificationsStore and its test

itisAliRH committed:
remove notification favorite field from backend and update schema, migration and tests

itisAliRH committed:
Merge branch 'dev' into notification-system

itisAliRH committed:
remove favorite query check on cleanup_expired_notifications in notifications manager

itisAliRH committed:
fix ActivityItem test error

itisAliRH committed:
python lint

jdavcs committed:
Do not start a new db transaction

jdavcs committed:
Update schema.ts (remove mention of dropped favorite attr)

jdavcs committed:
Fix logic in cleanup method

- Fix error: notification records are now deleted.

- Now that we don't have a `favorite` attribute, all that complicated
  logic is no longer necessary.

- We don't need to synchronize with the session becase the session is
  expired at the end of this method.

jdavcs reviewed (approved):
looks great!
